src/VisualStudio/Next/FindReferences/StreamingFindReferencesPresenter.TableDataSourceFindReferencesContext.cs

```diff
@@ -358,22 +358,132 @@ private TextSpan GetRegionSpanForReference(SourceText sourceText, TextSpan refer
-                    semanticModel, widenedSpan, document.Project.Solution.Workspace,
-                    insertSourceTextInGaps: true,
-                    cancellationToken: cancellationToken).ConfigureAwait(false);
+                await classificationService.AddSyntacticClassificationsAsync(
+                    document, widenedSpan, syntaxSpans, cancellationToken).ConfigureAwait(false);
+                await classificationService.AddSemanticClassificationsAsync(
+                    document, widenedSpan, semanticSpans, cancellationToken).ConfigureAwait(false);
 
-                var taggedText = classifiedLineParts.ToTaggedText();
+                var allParts = MergeClassifiedSpans(
+                    syntaxSpans, semanticSpans, widenedSpan, sourceText);
+                var taggedText = allParts.ToTaggedText();
 
                 var highlightSpan = new TextSpan(
-                    start: referenceSpan.Start - widenedSpan.Start, 
+                    start: referenceSpan.Start - widenedSpan.Start,
                     length: referenceSpan.Length);
 
                 return new TaggedTextAndHighlightSpan(taggedText, highlightSpan);
             }
 
+            private List<SymbolDisplayPart> MergeClassifiedSpans(
+                List<ClassifiedSpan> syntaxSpans, List<ClassifiedSpan> semanticSpans, 
+                TextSpan widenedSpan, SourceText sourceText)
+            {
+                // The spans produced by the language services may not be ordered
+                // (indeed, this happens with semantic classification as different
+                // providers produce different results in an arbitrary order).  Order
+                // them first before proceeding.
+                syntaxSpans = Order(syntaxSpans);
+                semanticSpans = Order(semanticSpans);
+
+                // Produce SymbolDisplayParts for both sets of ClassifiedSpans.  This will
+                // also produce parts for the regions between the sections that the classifiers
+                // returned results for (i.e. for things like spaces and plain text).
+                var syntaxParts = Classifier.ConvertClassifications(
+                    sourceText, widenedSpan.Start, syntaxSpans, insertSourceTextInGaps: true);
+                var semanticParts = Classifier.ConvertClassifications(
+                    sourceText, widenedSpan.Start, semanticSpans, insertSourceTextInGaps: true);
+
+                // Now merge the lists together, taking all the results from syntaxParts
+                // unless they were overridden by results in semanticParts.
+                return CreateAllParts(syntaxParts, semanticParts);
+            }
+
+            private List<ClassifiedSpan> Order(List<ClassifiedSpan> syntaxSpans)
+            {
+                return syntaxSpans.OrderBy(s => s.TextSpan.Start).ToList();
+            }
+
+            private List<SymbolDisplayPart> CreateAllParts(
```

Comment:
Maybe `MergeParts` instead of `CreateAllParts`? You're re-using existing parts, not creating new ones.